modules/README.md

```diff
@@ -1,58 +1,164 @@
+# Module requirements
 
-# PXP agent architecture
+### Metadata
 
-map of name:module, possibly factory style from peg
+The `metadata` of a module is a JSON object that contains a number of JSON
+schemas for specifying configuration options and actions. It contains:
 
-Module abstraction
+ - **configuration**: (optional) schema that describes the module configuration format;
+ - **actions**: an array where each item is an object that describes an action implemented by the module (please, refer to the below schema).
 
-    class Module {
-        std::string name;
-        bool invoke(const CthunMessage& request, CthunMessage& response);
-        Schema actions;
+The `metadata` schema is:
+
+    {
+        "type" : "object",
+        "description" : "Metadata format",
+        "properties" : {
+            "actions" : {
+                "type" : "array",
+                "items" : {
+                    "type" : "object",
+                    "description" : "Describes an action",
+                    "properties" : {
+                        "name" : {
+                            "type" : "string",
+                            "description" : "Name of the action",
+                        },
+                        "input" : {
+                            "type" : "object",
+                            "description" : "Schema for the PXP request arguments",
+                        },
+                        "results" : {
+                            "type" : "object",
+                            "description" : "Schema for the action results",
+                        },
+                    },
+                    "required" : ["description", name", "input", "results"],
```

Comment:
Maybe I'm still not getting it, but it feels like you should have the ```description``` property along the ```name```, ```input``` & ```output```:
```JSON
                    "properties" : {
                         "name" : {
                             "type" : "string",
                             "description" : "Name of the action",
                         },
                         "description" : {
                             "type" : "string",
                             "description" : "Description of the action",
                         },
                         "input" : {
                             "type" : "object",
                             "description" : "Schema for the PXP request arguments",
                         },
                         "results" : {
                             "type" : "object",
                             "description" : "Schema for the action results",
                         },
                     },
```